public/framework/catalog_cache/cache_loader.tsx

```diff
@@ -0,0 +1,278 @@
+  let query;
+  switch (loadCacheType) {
+    case 'databases':
+      query = `SHOW SCHEMAS IN ${addBackticksIfNeeded(dataSourceName)}`;
+      break;
+    case 'tables':
+      query = `SHOW TABLES IN ${addBackticksIfNeeded(dataSourceName)}.${addBackticksIfNeeded(
+        databaseName!
+      )}`;
+      break;
+    case 'accelerations':
+      query = `SHOW FLINT INDEX in ${addBackticksIfNeeded(dataSourceName)}`;
+      break;
+    default:
+      query = '';
+      break;
+  }
+  return query;
+};
+
+export const useLoadToCache = (loadCacheType: LoadCacheType) => {
+  const sqlService = new SQLService(coreRefs.http!);
+  const [currentDataSourceName, setCurrentDataSourceName] = useState('');
+  const [currentDatabaseName, setCurrentDatabaseName] = useState<string | undefined>('');
+  const [loadStatus, setLoadStatus] = useState<DirectQueryLoadingStatus>(
+    DirectQueryLoadingStatus.SCHEDULED
+  );
+
+  const {
+    data: pollingResult,
+    loading: _pollingLoading,
+    error: pollingError,
+    startPolling,
+    stopPolling: stopLoading,
+  } = usePolling<any, any>((params) => {
+    return sqlService.fetchWithJobId(params);
+  }, ASYNC_POLLING_INTERVAL);
+
+  const startLoading = (dataSourceName: string, databaseName?: string) => {
+    setCurrentDataSourceName(dataSourceName);
+    setCurrentDatabaseName(databaseName);
+
+    let requestPayload: DirectQueryRequest = {
+      lang: 'sql',
+      query: createLoadQuery(loadCacheType, dataSourceName, databaseName),
+      datasource: dataSourceName,
+    };
+
+    const sessionId = getAsyncSessionId(dataSourceName);
+    if (sessionId) {
+      requestPayload = { ...requestPayload, sessionId };
+    }
+
+    sqlService
+      .fetch(requestPayload)
+      .then((result) => {
+        setAsyncSessionId(dataSourceName, getObjValue(result, 'sessionId', null));
+        if (result.queryId) {
+          startPolling({
+            queryId: result.queryId,
+          });
+        } else {
+          console.error('No query id found in response');
+          setLoadStatus(DirectQueryLoadingStatus.FAILED);
+          updateToCache(null, loadCacheType, currentDataSourceName, currentDatabaseName);
+        }
+      })
+      .catch((e) => {
+        setLoadStatus(DirectQueryLoadingStatus.FAILED);
+        updateToCache(null, loadCacheType, currentDataSourceName, currentDatabaseName);
+        const formattedError = formatError(
+          '',
+          'The query failed to execute and the operation could not be complete.',
+          e.body.message
+        );
+        coreRefs.core?.notifications.toasts.addError(formattedError, {
+          title: 'Query Failed',
+        });
+        console.error(e);
+      });
+  };
+
+  useEffect(() => {
+    // cancel direct query
+    if (!pollingResult) return;
+    const { status: anyCaseStatus, datarows, error } = pollingResult;
+    const status = anyCaseStatus?.toLowerCase();
+
+    if (status === DirectQueryLoadingStatus.SUCCESS || datarows) {
```

Comment:
I see you're reusing some of the polling logic here, I'm wondering if its significantly better to abstract the actual polling+handling of responses such that both the cache loader and the direct query search don't have to deal with polling-specific details. If need be, we can open an issue and I can do this refactoring.

Reply:
Maybe there should be a more general wrapper for polling? I also have written some manual polling logic elsewhere.

Reply:
I totally agree, I'm planning to create hook for this in my next PR for create acceleration flyout (will surely re-use it here and refactor).

Reply:
Created an issue here: https://github.com/opensearch-project/dashboards-observability/issues/1503